driver/utils/unicode_converter.h

```diff
@@ -0,0 +1,722 @@
+
+#include <string>
+#include <string_view>
+#include <vector>
+
+using ConverterPivotWideCharType = UChar;
+inline const std::string converter_pivot_wide_char_encoding = "UTF-16";
+
+template <typename CharType>
+inline auto make_raw_str(std::initializer_list<CharType> && list) {
+    return std::string{list.begin(), list.end()};
+}
+
+inline auto sameEncoding(const std::string & lhs, const std::string & rhs) {
+    return (ucnv_compareNames(lhs.c_str(), rhs.c_str()) == 0);
+}
+
+// Return the same string with signature prefix removed, if it matches the provided signature/BOM byte array.
+// Even though the signature is specified in bytes, it will be considered matching only if it matches up to some character boundary.
+template <typename CharType>
+inline auto consumeSignature(std::basic_string_view<CharType> str, const std::string_view & signature) {
+    if (
+        signature.size() > 0 &&
+        (signature.size() % sizeof(CharType)) == 0 &&
+        (str.size() * sizeof(CharType)) >= signature.size() &&
+        std::memcmp(&str[0], &signature[0], signature.size()) == 0
+    ) {
+        str.remove_prefix(signature.size() / sizeof(CharType));
+    }
+
+    return str;
+}
+
+template <typename CharType>
+inline std::size_t consumeSignatureInPlace(std::basic_string<CharType> & str, const std::string_view & signature) {
+    const auto str_v_no_sig = consumeSignature(make_string_view(str), signature);
+    const auto symbols_to_trim = str.size() - str_v_no_sig.size();
+
+    if (symbols_to_trim > 0) {
+        // Doing this manually to guarantee that buffer pointers are not invalidated by the implementation.
+        const auto new_size = str.size() - symbols_to_trim;
+        std::memmove(&str[0], &str[symbols_to_trim], new_size * sizeof(CharType));
+        str.resize(new_size);
+    }
+
+    return symbols_to_trim; // Already trimmed symbols count, actually.
+}
+
+class UnicodeConverter;
+
+template <typename SourceCharType, typename DestinationCharType, typename PivotCharType>
+inline void convertEncoding(
+    UnicodeConverter & src_converter, const std::basic_string_view<SourceCharType> & src,
+    std::basic_string<PivotCharType> & pivot,
+    UnicodeConverter & dest_converter, std::basic_string<DestinationCharType> & dest,
+    const bool ensure_src_signature = true,
+    const bool trim_dest_signature = true
+);
+
+template <typename SourceCharType, typename DestinationCharType, typename PivotCharType>
+inline void convertEncoding(
+    UnicodeConverter & src_converter, const std::basic_string<SourceCharType> & src,
+    std::basic_string<PivotCharType> & pivot,
+    UnicodeConverter & dest_converter, std::basic_string<DestinationCharType> & dest,
+    const bool ensure_src_signature = true,
+    const bool trim_dest_signature = true
+) {
+    return convertEncoding(src_converter, make_string_view(src), pivot, dest_converter, dest, ensure_src_signature, trim_dest_signature);
+}
+
+class UnicodeConverter {
```

Comment:
I believe there should be a class-level comment, describing what does it converts and to what.

Reply:
Done.